[Woo POS] Centring of views

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13701 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR fixes some issues with incorrect vertical positioning of certain views in the POS, compared to designs.

1. Product error and empty views are not properly centered

| Design | App (old) | App (new) |
|--------|----------|------------|
| ![image](https://github.com/user-attachments/assets/5b6a884b-ecf9-4adb-99ae-a14637721506) | ![image](https://github.com/user-attachments/assets/6c222781-d9f9-42c0-ac4d-2a64985931c9) | ![products load error centring](https://github.com/user-attachments/assets/8729687a-e4b3-4df6-9a39-98718fcde9b7) |
| ![image](https://github.com/user-attachments/assets/e8a7615e-3475-47e8-af26-abcd24c49fab) | ![Simulator Screenshot - iPad Air 11-inch (M2) - 2024-08-22 at 14 29 07](https://github.com/user-attachments/assets/b8f216f0-56b9-46e2-9ab9-fd2b091ed706) | ![products empty state centring](https://github.com/user-attachments/assets/a17f2452-1a51-4960-839e-39760127794f) |

N.B. I also put the correct image in for the magnifier. We already have a very similar image in the main app, but it's a darker grey for POS, and since it's two toned I can't just use it as a template.

2. Cart placeholder 

| Design | App (old) | App (new) |
|--------|----------|------------|
| ![Image](https://github.com/user-attachments/assets/02016dc2-ebb4-4a80-a960-095997cd8116) | ![Image](https://github.com/user-attachments/assets/d78652af-d161-49e0-a818-1fcc961930f2) | ![cart empty state centring](https://github.com/user-attachments/assets/4caa6a3b-886c-4669-872e-5696f7a7c63c) |

3. Payment views

| Design | App (old) | App (new) |
|--------|----------|------------|
| ![Image](https://github.com/user-attachments/assets/4ca8b300-4fb3-4e38-af4b-5520c9f7a8c1) | ![Image](https://github.com/user-attachments/assets/5a3e1817-e834-4edb-869e-63a9611d9d33) | ![payment error](https://github.com/user-attachments/assets/88165a95-cbc3-41a0-82f5-222d10c8cfc1) |


## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Set up Proxyman or Charles to have a breakpoint on the response to `https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/*/rest-api/?json=true&path=/wc/v3/products*` and proxy your device's connection

### Products load error

1. Launch the app and navigate to POS
2. Abort the first request (and any retries)
3. Observe that the payments error state matches designs – i.e. centred vertically on the screen, not between other views.

### Products empty state

1. Continue from above
2. Tap Retry
3. Delete the content of the `data` array in the response
4. Execute the edited response
5. Observe that the empty state matches designs – i.e. centred vertically on the screen, not between the cart and foot of the screen.

### Cart empty state

Turn off your breakpoints

1. Exit POS, and enter again
2. Observe that the empty cart shows, and the positioning matches designs – i.e. the text is centred vertically on the screen, and the image is ~40 points above it.

N.B. Strictly, the design is centred between the header and the invisible button for Check out, but that is essentially the same as centred vertically on screen. Also, the way I've done it means I can't be totally precise for the space between the image and the text, but this was the most sensible way I could find to centre the text and then position the image relative to it.

### Fullscreen payments errors

1. Add a product to your cart with an error amount, e.g. ending in .01
2. Connect your reader
3. Go through to checkout
4. Tap your card when prompted
5. Observe that the error is shown with 120px padding from the top of the screen, not vertically centred.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

It's good to test the surrounding views especially for payments. I checked all the states I could think of, and also tested portrait mode and split view, such as they are.

No unit tests for this UI work.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/70291cda-14df-4f44-ab62-7327d6e3f8be

---

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
13701 Update cart empty state vertical positioning

joshheald committed:
13701 Vertically centre products error content

The `Products` title is sort-of background information and shouldn’t push the error content down.

joshheald committed:
13701 Vertically centre empty products view

We use the same logic to disregard the `Products` header in our vertical centring for both empty and error states.

joshheald committed:
Use correct image for empty POS product view

joshheald committed:
13701 Show fullscreen payment errors top aligned

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/221">20.2</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -224,6 +217,29 @@ private extension CartView {
+            Spacer()
+            // By designs, the text should be vertically centred with the image 40px above it.
+            // SwiftUI doesn't allow us to absolutely pin a view to the centre then position other views relative to it
+            // Instead, we can centre the text, and then put the image in an offset overlay.
```
> // Instead, we can centre the text, and then put the image in an offset overlay.

That's a smart approach. Unfortunately, it doesn't quite support larger dynamic-type sizes.

![image](https://github.com/user-attachments/assets/aeb14a1d-d709-4ec8-8e14-f8cb6dd36113)

staskus replied:
Using overlay looks to be a good option but it doesn't give us that control of layout. This is one of the things that would be easy to do with `UIKit` constraints but seems to be much harder to do with `SwiftUI` 🤔 I'm interested how to achieve it nicely as well. I'll let you know if I encounter a good solution.

joshheald replied:
Dangit. Thanks for looking further and finding a fix, I'll put that in

staskus commented on WooCommerce/Classes/POS/Presentation/TotalsView.swift:
```diff
@@ -267,15 +267,31 @@ private extension TotalsView {
             bottomPadding: 40
         )
+
+        static let topAligned = CardReaderViewLayout(
```
I admit seeing a non-centered payment error view looked weird to me at first 😀.

Figma comment mentioned that it was done to make sure that there's always space for "No internet connection" banner.

Either way, happy that `CardReaderViewLayout` has proven to be useful in one more case.

joshheald replied:
Yeah I was dreading this one when I saw how the code was written, then really happy to come across the Layout ❤️

staskus reviewed (approved):
✅ Thanks! Solid work.

I tested all the changed views: product error, product empty, cart empty, payment error in different configurations.

The only issue I found is with Cart + Dynamic Type.

staskus commented:
Another question.

Does the payment view usually chop like that for you or does it only appear in the video, @joshheald ?

<img width="814" alt="image" src="https://github.com/user-attachments/assets/baedbfff-54cb-4af4-b545-5fcb665f935a">

staskus commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -224,6 +217,29 @@ private extension CartView {
+                .lineSpacing(Constants.cartEmptyViewTextLineHeight)
+                .foregroundColor(Color.posTertiaryText)
+                .multilineTextAlignment(.center)
+                .overlay(alignment: .center) {
```
Turns out dynamic type issue is easy solvable by aligning this overlay on `top`.

```suggestion
                .overlay(alignment: .top) {
```

joshheald commented:
> Another question.
> 
> Does the payment view usually chop like that for you or does it only appear in the video, @joshheald ?

Thanks for the reminder – yes, I noticed that and wanted to check trunk to see whether it's introduced here, somehow. I don't see how it would be, but hey, transitions are fickle.

joshheald committed:
13701 Prevent shopping bags overlapping with text

At larger dynamic text sizes, the overlay would not be offset enough from the center to clear the text, and they would intersect.

Offsetting from the `.top` avoids this issue, since the image is above the text.

joshheald commented:
@staskus the cut between `Remove card` and `Processing` happens for me at 3845628, the commit on trunk that I branched from, and the one where I merged the processing animation. So I'm going to merge this PR, because it's unrelated to these changes.

It still happens on latest trunk too. I'll take a look at it separately, it's probably from my last PR.
